aten/src/ATen/native/sparse/utils/ParamUtils.h

```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+#include <ATen/ATen.h>
+#include <ATen/TensorUtils.h>
+#include <tuple>
+
+namespace at {
+namespace native {
+
+namespace apply {
+
+std::pair<Tensor, Tensor> softmax_sparse_check_invariants(
+    const Tensor& input_,
+    const int64_t dim_,
+    const bool half_to_float,
+    CheckedFrom function_name) {
+  TORCH_INTERNAL_ASSERT(input_.is_sparse());
+  TORCH_CHECK(
+      !half_to_float,
+      std::string(function_name) +
+          ": with half to float conversion is not supported on CPU");
```

Comment:
this references the CPU but is checked by both CPU and CUDA, right? Maybe "... half to float conversion is not supported for sparse inputs"?

Reply:
Yes you are right, so now it is:
`TORCH_CHECK(!half_to_float, std::string(function_name) + ": with half to float conversion is not supported on " + input_.device().str());`

About last question, half_to_float is not supported even with dense tensors, see here: https://github.com/pytorch/pytorch/blob/bcb0bb7e0e03b386ad837015faba6b4b16e3bfb9/aten/src/ATen/native/SoftMax.cpp#L121
I am now sure why.